executor/docker_executor.go

```diff
@@ -0,0 +1,147 @@
+	TimeLimits         []int
+	MemoryLimits       []int
+}
+
+type ExecutionResult struct {
+	ExitCode int
+	ExecTime float64
+}
+
+type DockerExecutor struct {
+	cli *client.Client
+}
+
+func NewDockerExecutor() (*DockerExecutor, error) {
+	cli, err := client.NewClientWithOpts(
+		client.FromEnv,
+		client.WithAPIVersionNegotiation(),
+	)
+	if err != nil {
+		return nil, err
+	}
+	return &DockerExecutor{cli: cli}, nil
+}
+
+func (d *DockerExecutor) ExecuteCommand(
+	solutionPath, messageID string, cfg CommandConfig,
+) error {
+	log := logger.NewNamedLogger("docker-executor")
+	ctx := context.Background()
+	start := time.Now()
+
+	// --- 1) Build the run_tests invocation ---
+	bin := filepath.Base(solutionPath) // e.g. "solution"
+	runCmd := fmt.Sprintf("run_tests.sh ./%s", bin)
+
+	// --- 2) Build the TIME_LIMITS and MEM_LIMITS env lines ---
+	const minMemKB = 128 * 1024 // 128 MB minimum per-test
+	timeEnv := make([]string, len(cfg.TimeLimits))
+	for i, s := range cfg.TimeLimits {
+		timeEnv[i] = strconv.Itoa(s)
+	}
+	memEnv := make([]string, len(cfg.MemoryLimits))
+	for i, kb := range cfg.MemoryLimits {
+		if kb < minMemKB {
+			kb = minMemKB
+		}
+		memEnv[i] = strconv.Itoa(kb)
+	}
+	env := []string{
+		"TIME_LIMITS=" + strings.Join(timeEnv, " "),
+		"MEM_LIMITS=" + strings.Join(memEnv, " "),
+		"INPUT_DIR=" + cfg.InputDirName,
+		"OUTPUT_DIR=" + cfg.OutputDirName,
+		"EXEC_RESULT_FILE=" + cfg.ExecResultFileName,
+	}
+
+	// --- 3) Container config: run from cfg.WorkspaceDir (under /tmp) ---
+	containerCfg := &container.Config{
+		Image:      cfg.DockerImage,
+		Cmd:        []string{"bash", "-lc", runCmd},
+		WorkingDir: cfg.WorkspaceDir,
+		Env:        env,
+		User:       "0:0",
+	}
+
+	// --- 4) Host config: mount the named volume over /tmp ---
+	hostCfg := &container.HostConfig{
+		AutoRemove:  true,
+		NetworkMode: container.NetworkMode("none"),
+		Resources: container.Resources{
+			PidsLimit: func(v int64) *int64 { return &v }(64),
+			Memory:    int64(maxIntSlice(cfg.MemoryLimits, 6*1024)) * 1024,
+			CPUPeriod: 100_000,
+			CPUQuota:  100_000,
+		},
+		Mounts: []mount.Mount{{
+			Type:   mount.TypeVolume,
+			Source: "worker_jobs-data",
+			Target: "/tmp",
+		}},
+		SecurityOpt:  []string{"no-new-privileges"},
+		CgroupnsMode: container.CgroupnsModePrivate,
+		IpcMode:      container.IpcMode("private"),
+	}
+
+	// --- 5) Create & Start ---
+	name := fmt.Sprintf("submission-%s", messageID)
+	resp, err := d.cli.ContainerCreate(ctx, containerCfg, hostCfg, nil, nil, name)
+	if err != nil {
+		log.Errorf("Create failed: %s [MsgID: %s]", err, messageID)
+		return err
+	}
+	if err := d.cli.ContainerStart(ctx, resp.ID, container.StartOptions{}); err != nil {
+		log.Errorf("Start failed: %s [MsgID: %s]", err, messageID)
+		return err
+	}
+	log.Infof("Started container %s [MsgID: %s]", resp.ID, messageID)
+
+	// --- 6) Wait for it to exit ---
+	statusCh, errCh := d.cli.ContainerWait(ctx, resp.ID, container.WaitConditionNotRunning)
```

Comment:
context used should be time limited, to timeout containers. In case of timeout, kill the container

Reply:
30 secs seems reasonable enough. Maybe even too much, but lets assume it takes some time to schedule a container